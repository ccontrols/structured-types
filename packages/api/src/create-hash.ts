import hash from 'object-hash';

export function createHash(
  name: string,
  content: Record<string, unknown>,
): string {
  return `${name}:${hash(content).slice(0, 10)}`;
}
