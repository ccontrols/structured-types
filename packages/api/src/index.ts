import * as ts from 'typescript';
import { getTypescriptConfig } from '@structured-types/typescript-config';
import { dirname } from 'path-browserify';
import { PropTypes, PropType, PropDiagnostic, FunctionProp } from './types';
import {
  tsDefaults,
  DocsOptions,
  ProgramOptions,
  getSymbolDeclaration,
} from './ts-utils';
import { SymbolParser } from './SymbolParser';
import { createHash } from './create-hash';

export * from './jsdoc';
export * from './ts';
export * from './types';
export * from './ts-utils';

const cleanPropParents = (
  prop: FunctionProp,
  propsName: 'properties' | 'parameters' | 'types',
  parents: Record<string, PropType>,
) => {
  if (prop[propsName]) {
    if (prop.parent && parents[prop.parent.name]) {
      delete prop[propsName];
    } else {
      prop[propsName] = consolidateParentProps(
        prop[propsName] as PropType[],
        parents,
      );
    }
  }
};
const consolidateParentProps = (
  props: PropType[],
  parents: Record<string, PropType>,
): PropType[] => {
  return props.map((prop) => {
    const propFn = prop as FunctionProp;
    cleanPropParents(propFn, 'properties', parents);
    cleanPropParents(propFn, 'parameters', parents);
    cleanPropParents(propFn, 'types', parents);
    return propFn;
  });
};

/**
 * API to analyze the given files
 * @example
 * import { analyzeFiles } from '@structured-types/api';
 *
 * const props = analyzeFiles(['index.ts'], {
 *  collectHelpers: true,
 *  collectSourceInfo: true,
 *  tsOptions: {
 *    allowJs: true,
 *  }
 * })
 * @param files list of files to be processed
 * @param options parsing options
 * @param programOptions typescript ts.program and ts.compilerHost
 * @returns the parsed types
 */
export const analyzeFiles = (
  files: string[],
  options: DocsOptions = {},
  programOptions: ProgramOptions = {},
): PropTypes => {
  const {
    tsOptions: userTsOptions,
    scope = 'exports',
    ...parseOptions
  } = options;
  const tsOptions = { ...tsDefaults, ...userTsOptions };
  const {
    extract,
    collectDiagnostics,
    internalTypes,
    collectHelpers,
    moduleCallback,
  } = parseOptions || {};
  const { program: userProgram, host: userHost, hostCallback } = programOptions;
  const host = userHost || ts.createCompilerHost(tsOptions);
  if (hostCallback) {
    hostCallback(host);
  }

  const program = userProgram || ts.createProgram(files, tsOptions, host);
  const defaultLibraryFileName = host.getDefaultLibFileName(tsOptions);

  const defaultLibraryPath = host.getDefaultLibLocation
    ? host.getDefaultLibLocation()
    : dirname(defaultLibraryFileName);
  const equalityComparer = host.useCaseSensitiveFileNames()
    ? (s1: string, s2: string) => s1 === s2
    : (s1: string, s2: string) => s1.toUpperCase() === s2.toUpperCase();

  const isLibraryFile =
    defaultLibraryPath === '/'
      ? program.isSourceFileDefaultLibrary
      : (file: ts.SourceFile, node: ts.Node) => {
          if (file.hasNoDefaultLib) {
            return true;
          }
          if (files.includes(file.fileName)) {
            return false;
          }
          if (typeof parseOptions.isInternal === 'function') {
            const result = parseOptions.isInternal(file, node);
            if (result !== undefined) {
              return result;
            }
          }
          return equalityComparer(dirname(file.fileName), defaultLibraryPath);
        };
  const parser = new SymbolParser(program, isLibraryFile, parseOptions);
  let parsed: PropTypes = {};
  const addSymbol = (symbol?: ts.Symbol): void => {
    if (symbol) {
      const symbolName = symbol.getName();
      if (
        internalTypes?.[symbolName] === undefined &&
        (!extract || extract.includes(symbolName))
      ) {
        const prop = parser.parse(symbol);
        if (prop) {
          parsed[symbolName] = prop;
        }
      }
    }
  };
  for (const fileName of files) {
    const sourceFile = program.getSourceFile(fileName);
    if (sourceFile) {
      if (scope === 'all') {
        ts.forEachChild(sourceFile, (node: ts.Node) => {
          const namedNode = node as ts.ClassDeclaration;
          if (namedNode.name) {
            const symbol = parser.checker.getSymbolAtLocation(namedNode.name);
            addSymbol(symbol);
          } else if (ts.isVariableStatement(node)) {
            node.declarationList.declarations.forEach((d) => {
              if (d.name) {
                const symbol = parser.checker.getSymbolAtLocation(d.name);
                addSymbol(symbol);
              }
            });
          }
        });
      }
      const module = parser.checker.getSymbolAtLocation(sourceFile);
      if (module) {
        const exports = parser.checker
          .getExportsOfModule(module)
          .sort((s1, s2) => {
            // getExportsOfModule returns unsorted exports
            const d1 = getSymbolDeclaration(s1);
            const d2 = getSymbolDeclaration(s2);
            return (
              (d1?.pos || Number.MAX_VALUE) - (d2?.pos || Number.MAX_VALUE)
            );
          });
        exports.forEach((symbol) => addSymbol(symbol));
        if (moduleCallback) {
          moduleCallback(module, parser.checker);
        }
      }
    }
  }
  if (extract?.length) {
    let propKeys = Object.keys(parsed);

    propKeys = propKeys.sort((key1, key2) => {
      return extract.indexOf(key1) - extract.indexOf(key2);
    });
    const sortedProps = propKeys.reduce(
      (acc, key) => ({
        ...acc,
        [key]: parsed[key],
      }),
      {},
    );
    parsed = sortedProps;
  }
  if (collectHelpers) {
    // only return parents that are not already exported from the same file
    const helpers: Record<string, PropType> = Object.keys(parser.helpers)
      .filter((helperName) => {
        const { name, token } = parser.helpers[helperName];

        if (options.collectHelpers && name) {
          const parsedNode = parsed[name];
          return (
            parsedNode === undefined ||
            createHash(name, {
              filePath: parsedNode.loc?.filePath,
              start: parsedNode.loc?.loc?.start,
              end: parsedNode.loc?.loc?.end,
            }) !== token
          );
        }

        return parsed[helperName] === undefined;
      })
      .reduce((acc, name) => ({ ...acc, [name]: parser.helpers[name] }), {});
    if (Object.keys(helpers).length) {
      parsed = Object.keys(parsed).reduce((acc, key) => {
        return {
          ...acc,
          [key]: consolidateParentProps([parsed[key]], helpers)[0],
        };
      }, {});

      parsed.__helpers = Object.keys(helpers).reduce((acc, key) => {
        return {
          ...acc,
          [key]: consolidateParentProps([helpers[key]], helpers)[0],
        };
      }, {});
    }
  }
  if (collectDiagnostics) {
    const allDiagnostics = ts
      .getPreEmitDiagnostics(program)
      .filter(({ file }) => file && files.includes(file.fileName));
    if (allDiagnostics.length) {
      parsed.__diagnostics = allDiagnostics.map(
        ({ category, messageText, file, start }) => {
          const message =
            typeof messageText === 'string'
              ? messageText
              : messageText.messageText;

          const result: PropDiagnostic = {
            category,
            message,
          };
          if (file) {
            result.fileName = file.fileName.split('\\').pop()?.split('/').pop();
            if (typeof start !== 'undefined') {
              const location = file.getLineAndCharacterOfPosition(start);
              result.row = location.line + 1;
              result.column = location.character + 1;
            }
          }
          return result;
        },
      );
    }
  }

  return parsed;
};

/**
 * API to analyze the given files by also loading the local typescript options from tsconfig
 * @example
 * import { parseFiles } from '@structured-types/api';
 *
 * const props = parseFiles(['index.ts'], {
 *  collectHelpers: true,
 *  collectSourceInfo: true,
 * })
 * @param files list of files to be processed
 * @param options parsing options
 * @param programOptions typescript ts.program and ts.compilerHost
 * @returns the parsed types
 */
export const parseFiles = (
  files: string[],
  options: DocsOptions = {},
  programOptions?: ProgramOptions,
): PropTypes => {
  if (!files.length) {
    throw new Error('You need to supply at least one file');
  }
  options.tsOptions = {
    ...tsDefaults,
    ...getTypescriptConfig(files[0], options.tsOptions, {
      host: programOptions?.host,
    }),
  };
  const results = analyzeFiles(files, options, programOptions);
  return results;
};
