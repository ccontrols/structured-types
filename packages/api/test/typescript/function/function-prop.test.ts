import path from 'path';
import { parseFiles } from '../../../src/index';

describe('function', () => {
  it('tuple parameters', () => {
    const results = parseFiles([path.resolve(__dirname, 'tuple-parameter.ts')]);
    expect(results).toEqual({
      distanceFromOrigin: {
        name: 'distanceFromOrigin',
        kind: 11,
        parameters: [
          {
            kind: 6,
            properties: [
              {
                kind: 2,
              },
              {
                kind: 2,
              },
            ],
          },
        ],
        returns: {
          kind: 2,
          optional: true,
        },
      },
    });
  });
  it('object-parameter', () => {
    const results = parseFiles([
      path.resolve(__dirname, 'object-parameter.ts'),
    ]);

    expect(results).toEqual({
      printCoord: {
        parameters: [
          {
            name: 'pt',
            description: 'object parameter',
            kind: 15,
            properties: [
              {
                description: 'x coordinate',
                kind: 2,
                name: 'x',
              },
              {
                description: 'optional y coordinate',
                kind: 2,
                optional: true,
                name: 'y',
              },
            ],
          },
        ],
        description: 'print coordinates',
        name: 'printCoord',
        kind: 11,
        returns: {
          kind: 12,
          optional: true,
        },
      },
    });
  });
  it('extends-parameter', () => {
    const results = parseFiles(
      [path.resolve(__dirname, 'extends-parameter.ts')],
      { collectHelpers: true },
    );
    expect(results).toEqual({
      paintHomeyBear: {
        name: 'paintHomeyBear',
        kind: 11,
        parameters: [
          {
            kind: 15,
            type: 'ExtendT',
            properties: [
              {
                name: 'm',
                parent: {
                  name: 'T',
                  token: 'T:27b151585c',
                },
                kind: 1,
                description: 'base type member property',
              },
              {
                name: 'honey',
                kind: 3,
                description: 'own member',
              },
            ],
            description: 'extended type',
          },
        ],
        returns: {
          kind: 14,
          type: 'Bear',
          extends: [{ name: 'Internal', token: 'Internal:1551f01e82' }],
          properties: [
            {
              name: 'honey',
              kind: 3,
              description: 'boolean type member',
            },
            {
              name: 'm',
              parent: {
                name: 'Internal',
                token: 'Internal:1551f01e82',
              },
              kind: 1,
              description: 'string type member',
            },
          ],
          description: 'interface extending another one',
          optional: true,
        },
        description: 'exported function',
      },
      __helpers: {
        'Bear:e62a2f1037': {
          description: 'interface extending another one',
          kind: 14,
          name: 'Bear',
          extends: [{ name: 'Internal', token: 'Internal:1551f01e82' }],
          properties: [
            {
              description: 'boolean type member',
              kind: 3,
              name: 'honey',
            },
            {
              description: 'string type member',
              kind: 1,
              name: 'm',
              parent: {
                name: 'Internal',
                token: 'Internal:1551f01e82',
              },
            },
          ],
          token: 'Bear:e62a2f1037',
        },
        'ExtendT:3552cd3c86': {
          description: 'extended type',
          kind: 15,
          name: 'ExtendT',
          properties: [
            {
              description: 'base type member property',
              kind: 1,
              name: 'm',
              parent: {
                name: 'T',
                token: 'T:27b151585c',
              },
            },
            {
              description: 'own member',
              kind: 3,
              name: 'honey',
            },
          ],
          token: 'ExtendT:3552cd3c86',
        },

        'T:27b151585c': {
          name: 'T',
          kind: 15,
          properties: [
            {
              name: 'm',
              kind: 1,
              description: 'base type member property',
            },
          ],
          description: 'base type',
          token: 'T:27b151585c',
        },

        'Internal:1551f01e82': {
          name: 'Internal',
          kind: 14,
          properties: [
            {
              name: 'm',
              kind: 1,
              description: 'string type member',
            },
          ],
          description: 'internal interface with one member',
          token: 'Internal:1551f01e82',
        },
      },
    });
  });
  it('inline-props', () => {
    const results = parseFiles([path.resolve(__dirname, 'inline-props.ts')], {
      collectHelpers: true,
    });

    expect(results).toEqual({
      inlineProps: {
        name: 'inlineProps',
        kind: 11,
        parameters: [
          {
            name: 'props',
            kind: 15,
            properties: [
              {
                name: 'stringProp',
                kind: 1,
              },
              {
                name: 'numProp',
                kind: 2,
              },
            ],
          },
        ],
        returns: {
          kind: 12,
          optional: true,
        },
      },
    });
  });
  it('class-prop', () => {
    const results = parseFiles([path.resolve(__dirname, 'class-prop.ts')]);
    expect(results).toEqual({
      Boz: {
        name: 'Boz',
        kind: 13,
        properties: [
          {
            name: 'fn',
            kind: 11,
            parameters: [
              {
                name: 'foo',
                kind: 13,
                type: 'Foo',
                properties: [
                  {
                    parent: {
                      name: 'Foo',
                      token: 'Foo:17cc78eebc',
                    },
                    static: true,
                    readonly: true,
                    name: 'dummy',
                    kind: 3,
                    value: false,
                  },
                ],
              },
            ],
            returns: {
              kind: 12,
              optional: true,
            },
          },
        ],
      },
    });
  });
  it('async-function', () => {
    const results = parseFiles([path.resolve(__dirname, 'async-function.ts')]);
    expect(results).toEqual({
      genMyClass: {
        async: true,
        name: 'genMyClass',
        kind: 11,
        returns: {
          description: 'Represents the completion of an asynchronous operation',
          kind: 14,
          name: 'Promise',
          type: 'PromiseConstructor',
          optional: true,
        },
      },
    });
  });

  it('generic-parameter', () => {
    const results = parseFiles(
      [path.resolve(__dirname, 'generic-parameter.ts')],
      { collectHelpers: true },
    );
    expect(results).toEqual({
      genericFunction: {
        name: 'genericFunction',
        kind: 11,
        parameters: [
          {
            name: 'box',
            kind: 14,
            generics: [
              {
                name: 'Type',
              },
            ],
            type: 'GenericInterface',
            properties: [
              {
                parent: {
                  name: 'GenericInterface',
                  token: 'GenericInterface:236d9b3683',
                },
                name: 'm',
                type: 'T',
              },
            ],
          },
          {
            name: 'newContents',
            type: 'Type',
          },
        ],
        returns: {
          kind: 14,
          generics: [
            {
              name: 'T',
            },
          ],
          type: 'GenericInterface',
          properties: [
            {
              name: 'm',
              type: 'T',
            },
          ],
          optional: true,
        },
        types: [
          {
            name: 'Type',
          },
        ],
      },
      __helpers: {
        'GenericInterface:236d9b3683': {
          name: 'GenericInterface',
          kind: 14,
          generics: [
            {
              name: 'T',
            },
          ],
          properties: [
            {
              name: 'm',
              type: 'T',
            },
          ],
          token: 'GenericInterface:236d9b3683',
        },
      },
    });
  });

  it('infer-return', () => {
    const results = parseFiles([path.resolve(__dirname, 'infer-return.ts')]);
    expect(results).toEqual({
      fn: {
        name: 'fn',
        kind: 11,
        parameters: [
          {
            name: 'in1',
            kind: 2,
          },
          {
            name: 'in2',
            kind: 2,
          },
        ],
        returns: {
          kind: 2,
          optional: true,
        },
      },
    });
  });
  it('react fc', () => {
    const results = parseFiles([path.resolve(__dirname, 'react-fc.ts')]);
    expect(results).toMatchSnapshot();
  });
  it('function-props', () => {
    const results = parseFiles([path.resolve(__dirname, 'function-props.ts')]);

    expect(results).toEqual({
      fn: {
        name: 'fn',
        kind: 11,
        properties: [
          {
            kind: 1,
            description: 'custom property for a function',
            name: 'customProp',
            value: 'my custom prop',
          },
        ],
        returns: {
          kind: 12,
          optional: true,
        },
      },
    });
  });

  it('spread-tuple-parameter', () => {
    const results = parseFiles([
      path.resolve(__dirname, 'spread-tuple-parameter.ts'),
    ]);
    expect(results).toEqual({
      spreadTupleFunction: {
        name: 'spreadTupleFunction',
        kind: 11,
        parameters: [
          {
            name: 'args',
            kind: 6,
            properties: [
              {
                kind: 1,
              },
              {
                kind: 2,
              },
              {
                kind: 7,
                prop: {
                  kind: 16,
                  properties: [
                    {
                      kind: 3,
                    },
                  ],
                },
              },
            ],
          },
        ],
        returns: {
          kind: 12,
          optional: true,
        },
      },
    });
  });

  it('union parameter', () => {
    const results = parseFiles([path.resolve(__dirname, 'union-parameter.ts')]);
    expect(results).toEqual({
      printId: {
        name: 'printId',
        kind: 11,
        parameters: [
          {
            name: 'id',
            kind: 4,
            properties: [
              {
                kind: 2,
              },
              {
                kind: 1,
              },
            ],
          },
        ],
        returns: {
          kind: 12,
          optional: true,
        },
      },
    });
  });

  it('jsdoc parameter', () => {
    const results = parseFiles([path.resolve(__dirname, 'jsdoc-parameter.ts')]);
    expect(results).toEqual({
      greet: {
        parameters: [
          {
            name: 'name',
            description: 'string type parameters',
            kind: 1,
          },
        ],
        description: 'greeting function',
        name: 'greet',
        kind: 11,
        returns: {
          kind: 12,
          optional: true,
        },
      },
    });
  });
  it('arrow fn jsdoc', () => {
    const results = parseFiles([path.resolve(__dirname, 'arrow-function.ts')]);
    expect(results).toEqual({
      arrowGreet: {
        description: 'arrow greeting function',
        name: 'arrowGreet',
        kind: 11,
        parameters: [
          {
            description: 'name parameter inline description',
            kind: 1,
            name: 'name',
          },
        ],
        returns: {
          kind: 12,
          optional: true,
        },
      },
    });
  });
});
