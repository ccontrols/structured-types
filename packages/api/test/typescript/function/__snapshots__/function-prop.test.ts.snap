// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`function react fc 1`] = `
Object {
  "TypedInitializedFunction": Object {
    "generics": Array [
      Object {
        "name": "Props",
      },
    ],
    "kind": 11,
    "name": "TypedInitializedFunction",
    "parameters": Array [
      Object {
        "generics": Array [
          Object {
            "name": "P",
          },
        ],
        "kind": 15,
        "name": "props",
        "properties": Array [
          Object {
            "kind": 1,
            "name": "m",
            "optional": true,
            "parent": Object {
              "name": "Props",
              "token": "f3f2f08a-587d-47f0-a905-613c94dc79d6",
            },
          },
          Object {
            "kind": 4,
            "name": "children",
            "optional": true,
            "parent": Object {
              "name": "PropsWithChildren",
              "token": "b2311635-2d2a-474e-9607-8e212ad22693",
            },
            "properties": Array [
              Object {
                "kind": 4,
                "properties": Array [
                  Object {
                    "kind": 4,
                    "properties": Array [
                      Object {
                        "generics": Array [
                          Object {
                            "name": "P",
                          },
                          Object {
                            "name": "T",
                          },
                        ],
                        "kind": 14,
                        "properties": Array [
                          Object {
                            "name": "type",
                          },
                          Object {
                            "name": "props",
                          },
                          Object {
                            "name": "key",
                          },
                        ],
                        "type": "ReactElement",
                      },
                      Object {
                        "kind": 4,
                        "properties": Array [
                          Object {
                            "kind": 1,
                          },
                          Object {
                            "kind": 2,
                          },
                        ],
                        "type": "ReactText",
                      },
                    ],
                    "type": "ReactChild",
                  },
                  Object {
                    "kind": 4,
                    "properties": Array [
                      Object {
                        "kind": 15,
                      },
                      Object {
                        "extends": Array [
                          Object {
                            "name": "ReadonlyArray",
                          },
                        ],
                        "kind": 14,
                        "properties": Array [
                          Object {
                            "index": Object {
                              "kind": 2,
                              "name": "n",
                            },
                            "kind": 20,
                            "prop": Object {
                              "type": "T",
                            },
                          },
                        ],
                        "type": "ReactNodeArray",
                      },
                    ],
                    "type": "ReactFragment",
                  },
                  Object {
                    "extends": Array [
                      Object {
                        "name": "ReactElement",
                        "token": "e47515cb-210c-418b-8682-c10b830f34a0",
                      },
                    ],
                    "kind": 14,
                    "properties": Array [
                      Object {
                        "kind": 4,
                        "name": "key",
                        "properties": Array [
                          Object {
                            "kind": 15,
                            "type": "Key",
                          },
                          Object {
                            "kind": 10,
                          },
                        ],
                      },
                      Object {
                        "kind": 15,
                        "name": "children",
                        "type": "ReactNode",
                      },
                      Object {
                        "name": "type",
                        "parent": Object {
                          "name": "ReactElement",
                          "token": "fc9257bd-9bc3-403a-a153-93112d896714",
                        },
                        "type": "T",
                      },
                      Object {
                        "kind": 17,
                        "name": "props",
                        "parent": Object {
                          "name": "ReactElement",
                          "token": "5b3ae18f-6315-4f73-bfac-10c52bf45142",
                        },
                        "type": "P",
                      },
                    ],
                    "type": "ReactPortal",
                  },
                  Object {
                    "kind": 3,
                  },
                  Object {
                    "kind": 10,
                  },
                  Object {
                    "kind": 8,
                  },
                ],
                "type": "ReactNode",
              },
              Object {
                "kind": 8,
              },
            ],
          },
        ],
        "type": "PropsWithChildren",
      },
    ],
    "properties": Array [
      Object {
        "kind": 4,
        "name": "propTypes",
        "optional": true,
        "parent": Object {
          "name": "FunctionComponent",
          "token": "ba3bf90e-4a0d-40ca-bd13-584249218941",
        },
        "properties": Array [
          Object {
            "generics": Array [
              Object {
                "name": "T",
              },
            ],
            "kind": 15,
            "type": "WeakValidationMap",
          },
          Object {
            "kind": 8,
          },
        ],
      },
      Object {
        "kind": 4,
        "name": "contextTypes",
        "optional": true,
        "parent": Object {
          "name": "FunctionComponent",
          "token": "f1c5edb2-3fd4-4ee5-9c29-fc1824946215",
        },
        "properties": Array [
          Object {
            "generics": Array [
              Object {
                "name": "T",
              },
            ],
            "kind": 15,
            "type": "ValidationMap",
          },
          Object {
            "kind": 8,
          },
        ],
      },
      Object {
        "kind": 4,
        "name": "defaultProps",
        "optional": true,
        "parent": Object {
          "name": "FunctionComponent",
          "token": "78508f56-e532-449b-8ba2-ee608fa3e1d8",
        },
        "properties": Array [
          Object {
            "generics": Array [
              Object {
                "type": "P",
              },
            ],
            "kind": 15,
            "type": "Partial",
          },
          Object {
            "kind": 8,
          },
        ],
      },
      Object {
        "kind": 4,
        "name": "displayName",
        "optional": true,
        "parent": Object {
          "name": "FunctionComponent",
          "token": "f8e74da5-c56f-4fab-bbe0-a66d77ebb71d",
        },
        "properties": Array [
          Object {
            "kind": 1,
          },
          Object {
            "kind": 8,
          },
        ],
      },
    ],
    "returns": Object {
      "kind": 12,
      "optional": true,
    },
    "type": "FC",
  },
}
`;
